### PR TITLE
[SPARK-30421][SQL] Dropped columns still available for filtering

### PR DESCRIPTION
**Pull request by amanomer**
### What changes were proposed in this pull request?
Prevent the filtering on dropped columns.

### Why are the changes needed?
Currently, analyzer rule `ResolveMissingReferences` adds the dropped column because drop is the partial implementation of Project and while applying rule it cannot distinguish between Project added by dropped. This PR reconstructs a dataframe for drop command.
See discussion in https://issues.apache.org/jira/browse/SPARK-30421

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by amanomer** [comment]
cc @maropu @cloud-fan @srowen

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -2376,7 +2376,18 @@ class Dataset[T] private[sql](
     if (remainingCols.size == allColumns.size) {
       toDF()
     } else {
-      this.select(remainingCols: _*)
+      val df = this.select(remainingCols: _*)
+      withAction("drop", df.queryExecution) { physicalPlan =>
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: why do we need to copy? and also `drop` isn't an action. Also, looks we're going to lose existing plans.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: e.g.:

```scala
Seq((1, 2), (3, 4)).toDF("a", "b").write.parquet("/tmp/abc")
spark.read.parquet("/tmp/abc").drop("a").explain(true)
```

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Probably, @amanomer just used my example code in the jira. That's just an example code and I didn't carefully check it (I just copied/pasted it from the other part). I'm not currently sure that this is an issue to fix as I said in the jira: https://issues.apache.org/jira/browse/SPARK-30421

**Comment by dongjoon-hyun** [comment]
ok to test

**Review by dongjoon-hyun — COMMENTED** [review]
There was some discussion on this on the JIRA. I want to add a note here because this follows Python `pandas` way.
```python
>>> import pandas as pd
>>> df = pd.DataFrame(data={'col1': [1, 2], 'col2': [3, 4]})
>>> df
   col1  col2
0     1     3
1     2     4
>>> df.drop(columns=["col1"]).loc[df["col1"] == 1]
   col2
0     3
```

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
@amanomer . The failure is repeated. It seems to be related.
> org.apache.spark.ml.classification.OneVsRestSuite.one-vs-rest: default params

**Comment by Dobiasd** [comment]
Not just this PR was closed, but also the [Jira issue](https://issues.apache.org/jira/browse/SPARK-30421) was resolved as "Won't Fix"? Could somebody please explain to me why? It the observed behavior intended, i.e., it's not a bug, it's a feature, or is it just not worth the effort to fix it?

To me, the below example still looks wrong.

```scala
scala> val df1 = Seq((0, "a"), (1, "b")).toDF("foo", "bar") 
df1: DataFrame = [foo: int, bar: string]

scala> val df2 = df1.drop("bar") 
df2: DataFrame = [foo: int]

scala> df2.printSchema 
root
 |-- foo: integer (nullable = false)

scala> df2.where($"bar" === "a").show 
+---+
|foo|
+---+
|  0|
+---+
```

Pandas, as a comparative example, behaves correctly:

```python
>>> import pandas as pd
>>> df1 = pd.DataFrame(data={'foo': [0, 1], 'bar': ["a", "b"]})
>>> df2 = df1.drop(columns=["bar"])
>>> df2.info()
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 1 columns):
foo    2 non-null int64
dtypes: int64(1)
memory usage: 144.0 bytes
>>> df2[df2["bar"] == "a"]
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/pandas/core/indexes/base.py", line 2897, in get_loc
    return self._engine.get_loc(key)
  File "pandas/_libs/index.pyx", line 107, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 131, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 1607, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 1614, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: 'bar'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py", line 2995, in __getitem__
    indexer = self.columns.get_loc(key)
  File "/usr/local/lib/python3.7/dist-packages/pandas/core/indexes/base.py", line 2899, in get_loc
    return self._engine.get_loc(self._maybe_cast_indexer(key))
  File "pandas/_libs/index.pyx", line 107, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 131, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 1607, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 1614, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: 'bar'
```

**Comment by cloud-fan** [comment]
It is a feature. If we drop this feature, we break other users' workloads.

**Comment by Dobiasd** [comment]
OK, thanks. Make sense.

Nevertheless, it somehow reminds me of the following. :wink:

workflow
(source: https://xkcd.com/1172/)

**Comment by srowen** [comment]
There isn't a 100% correct answer to a lot of these, but it seemed to be the consensus of several committers and the OP. "Because the software has worked this way" is indeed _an_ argument, but for me it's more compelling that "other similar software" works this way.

**Comment by Dobiasd** [comment]
For me, neither "Because the software has worked this way", nor "other similar software", are valid arguments. For me, it's just plainly wrong to be able to filter a dataframe on a column that does not exist in this dataframe.

I think this behavior is an issue because it means one can not simply look at the schema of a dataframe to determine if an operation with it is valid. Instead one has to consider the whole history of how the dataframe was created/derived. This leads to the effect that refactorings, e.g., changing the way of creation of a dataframe, will break one's code, even though the refactoring should be totally OK because it results in the exact same dataframe schema.

**Comment by jwdevs** [comment]
I've opened another one for this. perhaps it will give more argumentation. From my point of view this is a bug, not a feature... 
https://issues.apache.org/jira/browse/SPARK-45214

**Comment by cloud-fan** [comment]
Maybe we can add a config to make it strict, but the config should be off by default to not breaking existing workloads.

**Comment by jwdevs** [comment]
Perhaps it is a good compromise.

**Pull request closed without merging**